Proofread (#1)
IoannisNezis/ad-blog: content/post/project-qlue-ls/index.md

>(...) Tree-sitter \[can\] recognize many complete valid small fragments of a tree, but it might have trouble
> assembling them into incomplete larger fragments.

This is very useful for the use-case of syntax highlighting. You want to highlight as many tokens as possible, the larger context of in which they appear is often not so important.

In our use-case however, it's the other way around.

Let's look at a few examples.

## Implemented Capabilities

Okay, now that you may have an idea of the fundamental mechanisms.
Let's talk about the features, besides the lifecycle and synchronization, that I implemented.

{{< notice warning >}}
The implemented features are just a **proof of concept**!
There are many features that could and should be added.
But that takes a lot of time to get right.
And frankly also requires a stronger parser.
This should just give you an idea of what's possible.
{{< /notice >}}

### Formatting

When the client sends a `textDocument/formatting` request, the server returns a list of [`TextEdit`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textEdit)'s.

> A `TextEdit` is a incremental change, composed of a Range and a String.

Author: ```suggestion
A `TextEdit` is a incremental change, composed of a range and a string.
```
*hust*

> The Client then applies the changes and thus formats the document.

Author: ```suggestion
The client then applies the changes, thus formatting the document.
```
1. *husthust*
2. Gerunds (aka "-ing" Formen) sind generell ein besserer Stil.

I implemented this with a post-order-[traversal](https://en.wikipedia.org/wiki/Tree_traversal) of the parse-tree provided by tree-sitter.

![](img/postorder-traversal.png)
[^7]

> Each step the "Type" of the node is matched and handled with a strategy.

Author: ```suggestion
Each step, the "type" of the node is matched and handled with a strategy.
```
*husthusthust*

If you want to get a more detailed picture, check out the [formatting tests](https://github.com/IoannisNezis/Qlue-ls/blob/main/src/server/message_handler/formatting/tests.rs).

```sparql
BASE <http://...>
PREFIX namespace1: <iri>
PREFIX namespace12: <iri>
SELECT ?s ?p (MAX (?o)  AS ?max_a )
FROM <...>
FROM <...>
WHERE {
    {
        ?s namespace1:p1 ?o ;
          namespace12:p2 "dings" .
        ?other <abc> 32 .
        FILTER ((?other * 4) > 6 || ?o = "bar")
    }
    UNION {
        {
            SELECT * WHERE {
                ?a ?b ?c .
                BIND (?c + 3 AS ?var)
            }
            GROUP BY ?a
            HAVING ?b > 9
            ORDER BY DESC(?a)
            LIMIT 100
            OFFSET 10
        }
    }
}
```

```sparql
SELECT * WHERE {
    wd:Q11571 p:P166 [ pq:P585 ?date ] .
    wd:Q11572 p:P166 [
        pq:P585 ?date ;
        <...> ?other
    ]
    wd:Q11572 p:P166 []
}
```

```sparql
SELECT * WHERE {
    ?a ?b ",,," .
    FILTER (1 IN (1, 2, 3))
}
```

```sparql
SELECT * WHERE {
    ?a <iri>/^a/(!<>?)+ | (<iri> | ^a | a) ?b .
}
```

I added some optional formatting, like aligning prefix declarations and predicates.

> Here is an example for a query with every non default option:

Author: ```suggestion
Here is an example for a query with every non-default option:
```
non alleinstehend ist Französisch.

```sparql
prefix n1:     <...>
prefix n12:    <...>
prefix n123:   <...>
prefix n1234:  <...>
prefix n12345: <...>

select * 
where {
  ?var n1:p "ding" ;
       n12:p "foo" ;
       n123:p "bar" ;
       n1234:p "baz" ;
       n12345:p "out of filler" ;
}
```

Here is the full default format configuration:

```toml
[format]
align_predicates = true
align_prefixes = false
separate_prolouge = false
capitalize_keywords = true
insert_spaces = true
tab_size = 2
where_new_line = false
```

#### Ideas for the future

There are three things that the formatter does not handle well:

> - **Comments**: should be possible to place at the end of a line:

Author: ```suggestion
- **Comments**: should be able to be placed at the end of a line:
```
Die Kommentare selbst platzieren nichts am Ende der Zeile. Passiv, bitte.

    ```sparql
    SELECT * WHERE {
        ?s ?p ?o # EOL comment
    }
    ```

- **Long lines**: should cause a line break
    ```sparql
    SELECT ?v1 ?v3 ?v4 ?v5 ?v6 ?v7 ?v8 ?v9 ?v10
           ?v11 ?v12 ?v13 ?v14 ?v15 ?v16 ?v17
           ?v18 ?v19 ?v20 ?v21 ?v22 ?v23 ?v24
           ?v25 ?v26 ?v27 ?v28 ?v29 ?v30 ?v31
    WHERE {}
    ```

> - **small nested queries**: should be compressed

Author: Bitte entscheide dich, ob du am Anfang eines Stichpunkts groß- oder kleinschreiben willst und behalte es konsistent bei. Ich weiß, ich kann dasselbe mit meiner Redseligkeit nicht, aber trotzdem.

    ```sparql
    SELECT ?castle WHERE {
        osmrel:51701 ogc:contains ?castle .
        { { ?castle osmkey:historic "castle" }
          UNION
          { ?castle osmkey:historic "tower" . ?castle osmkey:castle_type "defensive" } }
        UNION
        { ?castle osmkey:historic "archaeological_site" . ?castle osmkey:site_type "fortification" }
        ?castle osmkey:name ?name .
        ?castle osmkey:ruins ?ruins .
        OPTIONAL { ?castle osmkey:historic ?class }
        OPTIONAL { ?castle osmkey:archaeological_site ?class }
    }
    ```

### Hover

> When the client sends a `textDocument/hover` to request, the server responds with information about a given position in a text-document.

Author: Sinn unklar - der Klient sendet ein hover zum request (das meint der Satz)??? Musst du selber nochmal formulieren oder mich wissen lassen, was genau du sagen willst.

This is useful to give inexperienced users documentation about how some constructs work:

 ![](img/examples/hover-filter.png)

#### Ideas for the future

> Here are a few ideas I have to extend this feature.

Author: Übersetzt heißt das ungefähr "Hier sind ein paar Ideen ich muss dieses Feature ausbauen." Wenn das kein Deutsch war, war deins kein Englisch. Was möchtest du hier sagen?
a) "Hier sind ein paar ideen. Ich muss dieses Feature ausbauen." (Klingt bissl wie BILD-Zeitungs-Deutsch, aber egal.)
b) "Hier sind ein paar Ideen, wie ich dieses Feature noch ausbauen kann." ("muss" würde in diesem Fall nicht passen. Pflichten sind keine Ideen.)
c) Hast du feature statt future geschreiben? Im Feature werden wir auch mit unseren Bachelors fertig. 

Ganz abgesehen davon:
```suggestion
Here are a few ideas I have to ameliorate this feature.
```
Ich schreibe hier etwas sehr formal (ist es schon gestelzt? Kann sein. Vielleicht ist auch "improve" die beste Wahl.), aber extend heißt meist physikalisch "ausstrecken" und meines Wissens nach geht das mit einem Feature nicht so gut.

**These are not implemented yet**

Display the structure of the query;
![](img/examples/hover-graph.png)
In a serious implementation, since these graph-patterns can get quiet complex, I would use [mermaid.js](https://mermaid.js.org/) to generate diagrams. This would require a custom plugin in the editor to render these diagrams.

> If the language server can query the knowledge graph autonomously (not implemented yet), it could display additional information retrieved from the knowledge graph.
> For example the label of a resource:

Author: ```suggestion
If the language server can query the knowledge graph autonomously (not implemented yet), it could display additional information retrieved from the knowledge graph, for example the label of a resource:
```
Das Thema mit den Drama-Punkten hatten wir schon.

The exemplary code action I implemented is "Shorten URI".
The idea is to shorten a URI into its compact ["Curie"](https://www.w3.org/TR/2010/NOTE-curie-20101216/) form.

| Before code action                       | After code action                       |
| ---------------------------------------- | --------------------------------------- |
| ![](img/examples/code-action-before.png) | ![](img/examples/code-action-after.png) |

This code-action is powered by [curies.rs](https://github.com/biopragmatics/curies.rs).

### Completion Suggestions
The key feature for a SPARQL language server is, in my opinion, code-completion.
Here the editor provides suggestions to the user.
In SPARQL this is not just a convenience. Without smart completion suggestsions a user has to know the ontology by heart. Its also a massive efficiency boost for experianced users.

Here is a possible structure for completions:

#### 1. Static

The simplest kind of completions is just to provide static snippets of keywords or structures to the user.
For this just the "Location" of the cursor is relevant, the context of the content of the knowledge-graph does not influence the suggestions.

> Here are a few Examples i implemented:

Author: ```suggestion
Here are a few examples i implemented:
```
*husthusthusthust*

|                                                |                                                |                                     |
| ---------------------------------------------- | ---------------------------------------------- | ----------------------------------- |
| ![](img/examples/cmp_select_pre.png)           | ![](img/examples/cmp_filter_pre.png)           | ![](img/examples/cmp_bind_pre.png)  |
| ![](img/examples/cmp_select_post.png)          | ![](img/examples/cmp_filter_post.png)          | ![](img/examples/cmp_bind_post.png) |

#### 2. Dynamic

Dynamic completions are more complex. They use all information available to provide context based suggestions.

This again could be split into two categories:

#### 2.1. Offline

> Here the Language server only uses "locally available" information:

Author: ```suggestion
Here, the language server only uses "locally available" information:
```
*husthusthusthusthust*

> That is: Text in the editor and the data bundled with the language server (for example known prefixes).

Author: ```suggestion
That is: text in the editor and the data bundled with the language server (for example known prefixes).
```
*husthusthusthusthusthust*
Ansonsten, ein Doppelpunkt nach nem Doppelpunkt ist nicht der beste Stil, aber wenns zum Tonfall zählt...

> A Simple example for this are suggestions of already defined variables:

Author: ```suggestion
A simple example for this are suggestions of already defined variables:
```
*husthusthusthusthusthusthust* scheint doch nicht so simple zu sein.

![](img/examples/cmp_variable.png)
This uses the parse-tree to find all variables.

> Currently, this is done very stupid, as this also suggests variables when they are out of scope:

Author: ```suggestion
Currently, this is done very stupidly, as this also suggests variables when they are out of scope:
```
Du hast stupid geschrieben. :D Mit grammatikalischer Korrektheit ist das keine Beleidigung.

![](img/examples/cmp_variable_dumb.png)

#### 2.1. Online

> Here the Language Server uses data from a SPARQL-endpoint to provide completion suggestions.

Author: ```suggestion
Here, the language server uses data from a SPARQL-endpoint to provide completion suggestions.
```
*husthusthusthusthusthusthusthust*

{{< notice warning >}}
This is not implemented yet!
{{< /notice >}}

> Here is a example from the Qlever-OSM-endpoint: <htps://qlever.cs.uni-freiburg.de/api/osm-planet/>.

Author: ```suggestion
Here is an example from the Qlever-OSM-endpoint: <htps://qlever.cs.uni-freiburg.de/api/osm-planet/>.
```
Nomen mit Vokal am Anfang immer mit "an".

[**O**pen**S**treet**M**ap](https://www.openstreetmap.org/) (OSM) is a Project that collects Geodata that is publicly accessible. Basically Google-maps, just without Google.
This data can be represented in an RDF knowledge graph and queried using SPARQL.
For example this query returns every bicycle parking spot in [Freiburg](https://www.openstreetmap.org/relation/62768):
```sparql
PREFIX geo: <http://www.opengis.net/ont/geosparql#>
PREFIX osmkey: <https://www.openstreetmap.org/wiki/Key:>
PREFIX ogc: <http://www.opengis.net/rdf#>
PREFIX osmrel: <https://www.openstreetmap.org/relation/>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?bicycle_parking ?geometry WHERE {
  osmrel:62768 ogc:sfContains ?bicycle_parking .
  ?bicycle_parking osmkey:amenity "bicycle_parking" ;
                   geo:hasGeometry/geo:asWKT ?geometry .
}
```

> Now I want to know the same thing just for Berlin.  

Author: ```suggestion
Now, I want to know the same thing, just for Berlin.  
```

> Unfortunately i forgot that the relation ID of Berlin is `62422`...  

Author: ```suggestion
Unfortunately, I forgot that the relation ID of Berlin is `62422`...  
```
...

> A good online-contextual-completion should help me:

Author: ```suggestion
A good online-contextual-completion can help me:
```
Das good online-contextual-completion muss überhaupt nichts. Kein guilt-tripping.

|                            |                           |
| -------------------------- | ------------------------- |
| ![](img/examples/online-cmp-before.png) | ![](img/examples/online-cmp-after.png) |

> # Using the Language server

Author: ```suggestion
# Using the language server
```
*husthusthusthusthusthusthusthusthust*

Ok, now let's talk about how to use the language server.
An editor needs to have a language server client and connect to our server.

I looked at three editors: neovim, vs-code, and a custom web-based-editor.

## Neovim

> To get a language server running with neovim is very easy because it has a build in language client and

Author: ```suggestion
To get a language server running with neovim is very easy because it has a built-in language client and
```
...nicht schon wieder...du hast es sogar angesprochen...

> neovim is build to be hacked.

Author: ```suggestion
neovim is built to be hacked.
```
...''...

### Installing the Program

> First `Qlue-ls` needs to be available as executable binary on the system, neovim runs on.

Author: ```suggestion
First ,`Qlue-ls` needs to be available as executable binary on the system neovim runs on.
```
Du brauchst beide gleichzeitig, oder? Oder soll qlue-ls auf ein System und Neovim läuft für sich weiter?

You could just build the binary from source.

> But to make it more convenient I added the binary into two "repositories":

Author: ```suggestion
But to make it more convenient, I added the binary into two "repositories":
```

The Rust repository [crate.io](https://crates.io/crates/qlue-ls). To install from there, run:
```shell
cargo install qlue-ls
```
And the python repository [Pypi](https://pypi.org/project/qlue-ls/). To install from there, run:
```shell
pipx install qlue-ls
```

> The python package is build with the help of [maturin](https://github.com/PyO3/maturin).

Author: ```suggestion
The python package is built with the help of [maturin](https://github.com/PyO3/maturin).
```
Ach man.

### Attaching

All you need is a `init.lua` file in your config directory with the following snippet:

```lua
vim.api.nvim_create_autocmd({ 'FileType' }, {
  desc = 'Connect to qlue-ls',
  pattern = { 'sparql' },
  callback = function()
    vim.lsp.start {
      name = 'qlue-ls',
      cmd = { 'qlue-ls', 'server' },
      root_dir = vim.fn.getcwd(),
      on_attach = function(client, bufnr)
        vim.keymap.set('n', '<leader>f', vim.lsp.buf.format, { buffer = bufnr, desc = 'LSP: ' .. '[F]ormat' })
      end,
    }
  end,
})
```

> When you open a `sparql` file (a file with suffix `.rq`).
> This runs the command `qlue-ls server` and connects the language client to this process.

Author: ```suggestion
When you open a `sparql` file (a file with suffix `.rq`), this runs the command `qlue-ls server` and connects the language client to this process.
```
No drama.

## VS-code

> In vs-code there is no build in language-client.

Author: ```suggestion
In vs-code, there is no built-in language-client.
```
Diesmal hätte ich hier das Schreien einführen sollen.

> Instead, you have to create a vs-code-extention that acts as a language-client.

Author: ```suggestion
Instead, you have to create a vs-code-extension that acts as a language-client.
```
Siehst du das Wort nicht tausendmal so oft wie ich?

> I will do that in the Future™.

Author: Ich denk mal Future^TM ist so gewollt.

## The Browser

> Okay, now the for the cherry on top:

Author: ```suggestion
Okay, now the cherry on top:
```
the for the

Let's connect this thing to a web-based editor (an editor that runs in a browser).

> "But wait!" you might say, "browser is JavaScript land".

Author: ```suggestion
"But wait!", you might say, "Browser is JavaScript land!".
```
Formulierung ist immer noch bisschen Kiezenglisch aber passt.

image: "img/cover.png"
---

> Modern developer environments are way more capable than simple text editors.

Author: t klein

Reviewer: oke
